### PR TITLE
Create optimised paint device to QgsGeometry converter

### PR DESCRIPTION
**Pull request by nyalldawson**
This is a step toward improving the performance of layout exports when label masking is enabled.

The intention here is to replace QgsMaskPaintDevice with a paint device which builds a QGIS geometry object from painted shapes, so that we can then use GEOS to perform the calculation of the entire clipping mask geometry instead of the equivalent QPainterPath methods. Proof of concept tests have demonstrated that using GEOS for the union/subtraction operations instead of the QPainterPath versions results in dramatic speed boosts when exporting with label masks.

### CODE REVIEW

**Comment by troopa81** [comment]
Don't we have an issue with the proposed solution because clipping geometry is not the same as clipping painted object?

clip_issue

**Comment by nyalldawson** [comment]
@troopa81 we'll still set the clipping mask on qpainter, but only after having calculated the overall merged mask using geos.

**Comment by troopa81** [comment]
> @troopa81 we'll still set the clipping mask on qpainter, but only after having calculated the overall merged mask using geos.

OK, my comment was based on [your proof of content](https://github.com/nyalldawson/QGIS/commit/4fa7c792be9f831c5036fab1d865aef0d0579356) and didn't see that this PR brings only the geomtrypaintdevice part.

Is it enough to solve the issue for project that crashes the application? Could you give us an hint on how much it reduces the generation time for a given project?

Could this not be fixed into Qt directly?

**Comment by nyalldawson** [comment]
@troopa81 

> My comment was based on your proof of content and didn't see that this PR brings only the geomtrypaintdevice part.

Ah, it's here: https://github.com/nyalldawson/QGIS/blob/4fa7c792be9f831c5036fab1d865aef0d0579356/src/core/symbology/qgssymbollayer.cpp#L983

I get the final dissolved mask via the geometry route, then convert to a painter path at the end to apply as the mask.

> Is it enough to solve the issue for project that crashes the application? 

Which project is that? (Right now I'm focusing solely on the slow exports to see if that can be improved)

> Could you give us an hint on how much it reduces the generation time for a given project?

Well, with hotspot I currently see the painter path internal merging occuring when subtracting the raw path as taking around 50% of the export time. With the (unoptimised!) change from the proof of concept branch the union/subtraction step is so fast it disappears from the profile entirely! 😁

> Could this not be fixed into Qt directly?

Theoretically yes, but (IMHO) practivally no. GEOS is designed for fast calculation of dissolving complex geometries and has been refined over decades. In contrast the qpainterpath internals are designed around simple paths with a small number of polygons and vertices.

**Comment by troopa81** [comment]
> I get the final dissolved mask via the geometry route, then convert to a painter path at the end to apply as the mask

OK with the proposed approach then. Why not completely rework QgsMaskPaintDevice/Engine to use geometry instead of adding a new geometry one? What would be the use of QgsMastPaintDevice/Engine?

> Which project is that? (Right now I'm focusing solely on the slow exports to see if that can be improved)

The one described in this issue should make it crash: https://github.com/qgis/QGIS/issues/50734

**Comment by nyalldawson** [comment]
> Why not completely rework QgsMaskPaintDevice/Engine to use geometry instead of adding a new geometry one?

That was my first approach, but the device class is exposed to python, so I can't do major reworking of that without breaking api. And QgsMaskPaintDevice/Engine are quite small classes, so I'd rather leave them around and mark as deprecated.

**Comment by nyalldawson** [comment]
> The one described in this issue should make it crash: #50734

Thanks, I'll use this to test!

**Comment by troopa81** [comment]
> And QgsMaskPaintDevice/Engine are quite small classes, so I'd rather leave them around and mark as deprecated.

OK to deprecated them.

> Thanks, I'll use this to test!

Really interested in the results because the crash seems to occur when calling [installing mask](https://github.com/qgis/QGIS/issues/50785#issue-1434866328), not really prepare them.

> Well, with hotspot I currently see the painter path internal merging occuring when subtracting the raw path as taking around 50% of the export time. 

How much is approximately export time in that situation? Merging/Substracting is suppose to happen only once per masked symbol layer, so I would not have bet that it was where the latency/crash issue was coming from (But maybe Qt is really really bad at it).

**Comment by nyalldawson** [comment]
@troopa81 

> Really interested in the results because the crash seems to occur when calling https://github.com/qgis/QGIS/issues/50785#issue-1434866328, not really prepare them.

Ok, I get that crash even with my proof of concept branch. Looks like the complex mask causes an overflow. It's happening in a few different places for me -- one frequent one is when it's serializing painter state and storing the masks.

This is why I believe the next optimisation would be to simplify the masks, eg by pre-clipping them to the feature extent (+a suitable buffer), and then simplifying to a reasonable level given export DPI. (I tested naively calling .simplify( 1 ) on the clip geometry from the POC branch, and the geometry size is reduced sufficiently to avoid the crash).

> How much is approximately export time in that situation? Merging/Substracting is suppose to happen only once per masked symbol layer, so I would not have bet that it was where the latency/crash issue was coming from (But maybe Qt is really really bad at it).

It is :rofl:  -- see https://github.com/qt/qtbase/blob/423ae2de96360619b5233ecb6a0be56b61ec58a3/src/gui/painting/qpathclipper.cpp#L1585. There's no indexing happening there, so the dissolve/clip is slooow, and gets much worse as more labels are added to the map.

Here's what hotspot shows for the export (master branch). The merging logic isn't the main chunk of time, but it is significant. It's my belief that the main hotspots could be addressed after we gain access to the geometry api on the masks, so that we can do all the fancy simplification and clipping logic to reduce the mask complexity, which we just don't have the flexibility to do with the painter paths:

image

**Comment by troopa81** [comment]
> Ok, I get that crash even with my proof of concept branch. Looks like the complex mask causes an overflow. It's happening in a few different places for me -- one frequent one is when it's serializing painter state and storing the masks.

That's what I add in mind. IIRC, Qt was enable to only keep a reference to the mask/clipPath, so for every masked feature, mask was added again in the pdf output, making it huge and crash the application.

> It is 🤣 -- see https://github.com/qt/qtbase/blob/423ae2de96360619b5233ecb6a0be56b61ec58a3/src/gui/painting/qpathclipper.cpp#L1585. There's no indexing happening there, so the dissolve/clip is slooow, and gets much worse as more labels are added to the map.

Indeed.

> This is why I believe the next optimisation would be to simplify the masks, eg by pre-clipping them to the feature extent (+a suitable buffer), and then simplifying to a reasonable level given export DPI. (I tested naively calling .simplify( 1 ) on the clip geometry from the POC branch, and the geometry size is reduced sufficiently to avoid the crash).

Simplifying mask seems to be a good things to do. But I don't think we should simplify according to DPI because when it comes to vector output user is supposed to be able to zoom in as much as he wants. But we could chose a reasonnable tolerance, or make it a user parameter.

Regarding the clipping to feature extent, it's not gonna simplify the clipPath, it just gonna reduce its extent. I would rather try to fix within Qt so it could understand that some clipPath are actually the same (using hash) and they should be referenced/linked not duplicated.

**Comment by nirvn** [comment]
I'm +1 to merge this. On the simplification front, I think we should handle that scenario the same way we do with vector layers when exporting to PDF (namely that we offer the users to simplify based on DPI or not simplify).

This optimized paint device could also have some positive impact on map canvas rendering, which would benefit both desktop as well as server.

File: src/core/painting/qgsgeometrypaintdevice.cpp [hunk 1]
```diff
@@ -0,0 +1,538 @@
...
+void QgsGeometryPaintEngine::drawPath( const QPainterPath &path )
+{
+  QPainterPath realPath = path;
+  if ( mUsePathStroker )
+  {
+    QPen pen = painter()->pen();
+    QPainterPathStroker stroker( pen );
+    QPainterPath strokedPath = stroker.createStroke( path );
+    realPath = strokedPath;
+  }
```
Comment (nirvn) [inline src/core/painting/qgsgeometrypaintdevice.cpp:467]: This is meant to not return and continue onward with the addSubpathGeometries() functio right?

Comment (nyalldawson) [reply src/core/painting/qgsgeometrypaintdevice.cpp:467]: Correct -- we stroke the path then fallback to the same logic for rendering the polygon path

**Review by nirvn — APPROVED** [review]

File: src/core/painting/qgsgeometrypaintdevice.cpp (hunk 1 above)
Comment (nirvn) [reply src/core/painting/qgsgeometrypaintdevice.cpp:467]: @nyalldawson , approved.

**Pull request merged**